content/blog/2022-05-19-CoC/index.md

```diff
@@ -0,0 +1,298 @@
+    $ t := x \; | \; \lambda x:t_1 \rightarrow t_2 \; | \; (t_1 \; t_2) \; | \; \forall x:t_1, t_2 \; | \; \text{Type} $
+</p>
+
+I will be using this syntax throughout the rest of the blog post to go over examples,
+but the inference rules remain the same as before.
+
+I wanted the ability to write proofs in my programs, so I designed the AST with
+this in mind:
+```
+type t = 
+  | Id of id
+  | Fun of id * t * t
+  | App of t * t
+  | Forall of id * t * t
+  | Type
+
+and theorem =
+{
+  id: id;
+  theorem: t;
+  proof: t;
+}
+
+and prog = 
+| Let of id * t * prog
+| Theorem of theorem * prog
+| Expr of t
+```
+
+Here is an example of a valid program
+```
+let Nat = forall N:Type, (forall _:N, (forall _:(forall _:N, N), N)) in
+let zero = lambda N:Type -> (lambda x:N -> (lambda _:(forall _:N, N) -> x)) in
+let succ = lambda nat:Nat -> (lambda N:Type -> (lambda x:N -> 
+  (lambda f:(forall _:N, N) -> (f (((nat N) x) f))))) in
+(succ (succ zero))
+(* output: (fun N:Type -> (fun x:N -> (fun f:(forall _:N , N) -> (f (f x))))) *)
+```
+and this is an example of a valid (unrelated) proof.
+```
+Theorem p_implies_p: forall P:Type, forall _:P, P.
+Proof. lambda P:Type -> lambda x:P -> x.
+```
+
+### Implementation
+
+The main implementation task was implementing a typechecker. To typecheck
+a term, we first check whether that term is a context. If it is, we typecheck
+the context to ensure that it is well-typed. Otherwise, the term is typechecked
+according to the typing rules above.
+
+Typechecking also needs to happen for entire programs. When typechecking
+a `Let` statement, beta reduction is performed on the term, and that result
+is substituted for the corresponding identifier throughout the rest of the program.
+Then the rest of the program is typechecked. When typechecking a `Theorem`,
+beta reduction is performed on the theorem term, and the proof term is typechecked.
+Then the typechecker checks if these two resulting terms are alpha equivalent.
+If they are, then the proof is valid. The proof term is then beta reduced
+and this term is substituted in for the corresponding identifier throughout the
+rest of the program. Then the rest of the program is typechecked.
+When the typechecker sees an `Expr`, it typechecks the
+term to make sure it is well-typed and then returns the beta-reduced term.
+
+## Evaulation
+
+My goal for this project was to be able to solve some of the simple exercises
+in the [Software Foundations](https://softwarefoundations.cis.upenn.edu/)
+logic chapter, and I was somewhat successful.
+I started by defining logical conjunction and disjunction.
+
+```
+(* Logical conjunction *)
+let conj = lambda P:Type -> lambda Q:Type -> forall x:Type, 
+    forall _:(forall _:P, forall _:Q, x), x in
+(* Logical disjunction *)
+let sum = lambda P:Type -> lambda Q:Type -> forall x:Type, 
+    forall _:(forall _:P, x), forall _:(forall _:Q, x), x in
+```
+
+Here are a few of the proofs I was able to successfully typecheck.
```

Comment:
Can you give a high-level synopsis of what these proofs are/where they came from? Are they from your own mind, or are they from SF somewhere?